core/include/vecmem/containers/vector.hpp

```diff
@@ -23,4 +26,20 @@ namespace vecmem {
     */
    template<typename T>
    using vector = std::vector<T, vecmem::polymorphic_allocator<T>>;
-}
+
+   /// Helper function creating a @c vecmem::device_vector_data object
+   template< typename TYPE, typename ALLOC >
+   VECMEM_HOST
+   device_vector_data< TYPE >
+   get_data( std::vector< TYPE, ALLOC >& vec );
```

Comment:
Same question as before, why do we want these to be class-external functions instead of static?

Reply:
Let me answer this one first...

Since `vecmem::vector` is (currently) just a specialisation of `std::vector`, we need to make `get_data(...)` external to it. Since C++ doesn't allow us to extend existing classes with additional functions. (I don't want to make `vecmem::vector` inherit from `std::vector`, as inheriting from STL types is generally not a good idea.)

Since for `vecmem::vector` we have to have `get_data(...)` as an external function, I thought it would be a nice design to just have `vecmem::get_data(...)` functions for all of the types that vecmem handles/supports. For instance I'm now playing with the following type:

```c++
namespace vecmem {
   template< typename T >
   using jagged_vector = vector< vector< T > >;
}
```

For which I'm introducing a special `vecmem::jagged_vector_data` type (I'll want to propose a change to the naming of the `vecmem::device_vector_data` types in an upcoming PR...), which would be created by another specialisation of the `vecmem::get_data(...)` function. :wink: